pkg/nsx/services/subnetipreservation/subnetipreservation.go

```diff
@@ -0,0 +1,192 @@
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/vmware-tanzu/nsx-operator/pkg/apis/vpc/v1alpha1"
+	"github.com/vmware-tanzu/nsx-operator/pkg/logger"
+	"github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/common"
+	nsxutil "github.com/vmware-tanzu/nsx-operator/pkg/nsx/util"
+)
+
+var (
+	log                             = &logger.Log
+	MarkedForDelete                 = true
+	ResourceTypeSubnetIPReservation = "DynamicIpAddressReservation"
+	ipReservationCRUIDIndexKey      = "ipReservationCRUID"
+	ipReservationCRNameIndexKey     = "ipReservationCRName"
+)
+
+type IPReservationService struct {
+	common.Service
+	IPReservationStore *IPReservationStore
+	builder            *common.PolicyTreeBuilder[*model.DynamicIpAddressReservation]
+}
+
+// InitializeService initializes SubnetIPReservationService service.
+func InitializeService(service common.Service) (*IPReservationService, error) {
+	builder, _ := common.PolicyPathVpcSubnetDynamicIPReservation.NewPolicyTreeBuilder()
+	wg := sync.WaitGroup{}
+	fatalErrors := make(chan error, 1)
+	defer close(fatalErrors)
+
+	ipReservationService := &IPReservationService{
+		Service:            service,
+		IPReservationStore: SetupStore(),
+		builder:            builder,
+	}
+
+	wg.Add(1)
+	go ipReservationService.InitializeIPReservationStore(&wg, fatalErrors)
+	wg.Wait()
+
+	if len(fatalErrors) > 0 {
+		err := <-fatalErrors
+		return ipReservationService, err
+	}
+
+	return ipReservationService, nil
+}
+
+// NSX does not implement search API for Subnet IPReservation.
+// InitializeIPReservationStore searches all the Subnets, lists the IPReservations
+// under those Subnets and saves the IPReservations created by the current cluster to store.
+func (s *IPReservationService) InitializeIPReservationStore(wg *sync.WaitGroup, fatalErrors chan error) {
+	defer wg.Done()
+	subnetStore := &SubnetStore{ResourceStore: common.ResourceStore{
+		Indexer:     cache.NewIndexer(keyFunc, cache.Indexers{}),
+		BindingType: model.VpcSubnetBindingType(),
+	}}
+	queryParam := fmt.Sprintf("%s:%s", common.ResourceType, common.ResourceTypeSubnet)
+	count, err := s.SearchResource(common.ResourceTypeSubnet, queryParam, subnetStore, nil)
+	if err != nil {
+		fatalErrors <- err
+		log.Error(err, "failed to get all NSX Subnets")
+		return
+	}
+	log.V(2).Info("Successfully fetch all Subnet from NSX", "count", count)
+	for _, obj := range subnetStore.List() {
+		subnet := obj.(*model.VpcSubnet)
+		subnetInfo, err := common.ParseVPCResourcePath(*subnet.Path)
+		if err != nil {
+			fatalErrors <- err
+			log.Error(err, "failed to parse Subnet path", "Subnet", *subnet.Path)
+			return
+		}
+		// TODO: support scale scenario
+		var cursor *string
+		pageSize := int64(1000)
+		markedForDelete := false
+		ipReservations, err := s.NSXClient.DynamicIPReservationsClient.List(subnetInfo.OrgID, subnetInfo.ProjectID, subnetInfo.VPCID, subnetInfo.ID, cursor, &markedForDelete, nil, &pageSize, nil, nil)
```

Comment:
a Question: if there are more than 1000 dynamic ip reservations on subnet, can this "List" return them all?